cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,612 @@
+        return self._normalize_variable(ret)
+
+    def generate_new_array(self, dtype, rshape, ashape, input_order=None):
+        """Generate new _FusionCudaArray object with a new memory space.
+        """
+        ret = self._generate_new_variable(
+            _FusionCudaArray,
+            dtype, rshape=rshape, ashape=ashape, input_order=input_order)
+        ret.memory.base_ashape = ret.ashape
+        return ret
+
+    def generate_new_scalar(self, dtype, **kwargs):
+        """Generate new _FusionCudaArray object with a new memory space.
+        """
+        return self._generate_new_variable(_FusionCudaScalar, dtype, **kwargs)
+
+    def make_view(self, var, **kwargs):
+        assert isinstance(var, _FusionCudaArray)
+        serial_number = self._serial_number
+        self._serial_number += 1
+        ret = var.make_view(serial_number, **kwargs)
+        return self._normalize_variable(ret)
+
+    def broadcast_to(self, var, ashape, rshape):
+        """Make a view of the input array with the given shape.
+        """
+        return self.make_view(
+            var, ashape=ashape, rshape=rshape, broadcasted_from=var)
+
+    def rotate_with_axis(self, var, axis):
+        """Make a view of an array by rotating ``var`` with given axis.
+        """
+        assert isinstance(var, _FusionCudaArray)
+        return self.make_view(var, rotated_from=var, axis=axis)
+
+    def index_with_int(self, var, key):
+        """Make a view of an array. by indexing ``var`` with given integer.
+        """
+        return self.index_with_tuple(var, (key,))
+
+    def index_with_tuple(self, var, key):
+        """Make a view of an array. by indexing ``var`` with given tuple.
+        """
+        for k in key:
+            if not isinstance(k, int):
+                raise IndexError(
+                    'Cannot subscript by type {}.'.format(type(k)))
+        if len(key) > var.ndim:
+            raise IndexError('too many indices for array.')
+
+        ashape = var.ashape[len(key):]
+        rshape = var.rshape[len(key):]
+
+        return self.make_view(
+            var, indexed_from=var, index_key=key,
+            ashape=ashape, rshape=rshape)
+
+    @property
+    def all_variables(self):
+        """Returns the list of all variables this class emitted.
+        """
+        return list(self._variables_dict.values())
+
+
+class _FusionHistory(object):
+    """History of operation exectuted in the target function of fusion.
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.vc = _VariableConductor()
+        self.shape_constraints = _fusion_shape._ShapeConstraints()
+        self.op_list = []
+        self.submodules = {}
+
+    @staticmethod
+    def _to_interface(x):
+        """Returns an _array or a _scalar object which packs the given value.
+        """
+        if isinstance(x, _fusion_variable._FusionCudaVarBase):
+            return x.as_interface()
+        if x is None:
```

Comment:
A nitpick but it's better to test this before `isinstance`.
And use `assert isinstance(...)` instead of `assert False`.